modules/tinymce/src/themes/silver/main/ts/ui/toolbar/button/ToolbarButtons.ts

```diff
@@ -79,6 +87,57 @@ const getTooltipAttributes = (tooltip: Optional<string>, providersBackstage: UiF
 
 const focusButtonEvent = Id.generate('focus-button');
 
+const renderCommonStructureUpdateableText = (
+  icon: Optional<string>,
+  text: Optional<Memento.MementoRecord>,
+  tooltip: Optional<string>,
+  receiver: Optional<string>,
+  behaviours: Optional<Behaviours>,
+  providersBackstage: UiFactoryBackstageProviders
+): AlloyButtonSpec => {
+  return {
+    dom: {
+      tag: 'button',
+      classes: [ ToolbarButtonClasses.Button ].concat(text.isSome() ? [ ToolbarButtonClasses.MatchWidth ] : []),
+      attributes: getTooltipAttributes(tooltip, providersBackstage)
+    },
+    components: componentRenderPipeline([
+      icon.map((iconName) => renderIconFromPack(iconName, providersBackstage.icons)),
+      text.map((mem) => mem.asSpec()),
+    ]),
+
+    eventOrder: {
+      [NativeEvents.mousedown()]: [
+        'focusing',
+        'alloy.base.behaviour',
+        'common-button-display-events'
+      ]
+    },
+
+    buttonBehaviours: Behaviour.derive(
+      [
+        DisablingConfigs.toolbarButton(providersBackstage.isDisabled),
+        ReadOnly.receivingConfig(),
+        AddEventsBehaviour.config('common-button-display-events', [
+          AlloyEvents.run<EventArgs<MouseEvent>>(NativeEvents.mousedown(), (button, se) => {
+            se.event.prevent();
+            AlloyTriggers.emit(button, focusButtonEvent);
+          })
+        ])
+      ].concat(
+        receiver.map((r) => Reflecting.config({
+          channel: r,
+          initialData: { icon, text },
+          renderComponents: (data: ButtonState, _state) => componentRenderPipeline([
+            data.icon.map((iconName) => renderIconFromPack(iconName, providersBackstage.icons)),
+            data.text.map((text) => renderLabel(text, ToolbarButtonClasses.Button, providersBackstage))
```

Comment:
I haven't looked at it properly yet, but at a glance, the thing that worries me is that in the UpdateMenuTextEvent handler, we are relying on the optMemDisplayText Memento to exist, but here on renderComponents, we don't keep that memento id. I'm actually not sure why you have `Reflecting` at all. And I think we probably don't want so much duplication between `renderCommonStructure` and `renderCommonStructureUpdateableText`.

I have a branch where I was cleaning up toolbar buttons because it's very messy and inconsistent (it was one section of the sink work that was shelved), and that might be the underlying problem - that the existing code is hard to follow. I'll have a proper look through this later, and maybe we can have a call about it so I can see what if there's a cleaner way to make this change.